compiler-rt/lib/radsan/radsan_interceptors.cpp

```diff
@@ -35,6 +36,12 @@ using namespace __sanitizer;
 
 namespace radsan {
 void expectNotRealtime(const char *intercepted_function_name) {
+  if (radsan::IsInitRunning())
+  {
+    return;
+  }
+  radsan::EnsureInitialized();
```

Comment:
Two questions here:

1. I don't yet understand why we need `EnsureInitialized` here - this expectation is only called from within an interceptor, isn't it? Doesn't that mean that `IsInitialized()` will always return `true` here?
2. If `EnsureInitialized()` checks `IsInitRunning` inside it, do we need to check it in the line above this one as well?

Reply:
So there are a few different branches here.

1. On linux machines, where we load the `radsan_init` into the `.preinit` array, this will always return true
2. On mac machines:

* The first thread going through here will initialize radsan
* any subsequent threads that have an interception happen during the "IsInitRunning" phase will fail that check above, and not register their interception
* After "IsInitRunning" is false, it will always be true.

This situation is a bit complex but I think it scratches all the itches of any corner cases (threads coming in at the same time, without things being initialized). I have to assume this is why msan did this in the first place.



One thing that would simplify all of this massively is: "Is there one single solitary point where we KNOW we will be able to initialize on one thread ONLY?"

To my knowledge, I can't think of anything on all systems (obviously `.preinit` exists on linux boxes). We have to use the first interception point on the first thread calling a `[[clang::realtime]]` to set ourselves up.